Add client login and basic error handling

SvantjeJung opened the pull request:
Added basic client functionality

SvantjeJung committed:
Add client login and basic error handling

SvantjeJung committed:
Add command functionality for client

Client can now send commands Quit, Ping and everything else as a Query.
But only Quit is working on client and server side. Other commands need
to be implemented in the future.

SvantjeJung committed:
Fixed debug messages on client disconnect in conn.rs

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,217 @@
+use uosql::net::Login;
+use uosql::net::Command;
+use byteorder::{ReadBytesExt, WriteBytesExt};
+use bincode::SizeLimit;
```
Those use statements can be reduced to:

``` rust
use std::io::{self, Write, Read, stdout};
use std::net::TcpStream;
use bincode::rustc_serialize::{decode_from, encode_into};
use uosql::net::{Cnv, Greeting, Login, Command};
use byteorder::{ReadBytesExt, WriteBytesExt};
use bincode::SizeLimit;
```

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,217 @@
+    match io::stdin().read_line(&mut input) {
+        Ok(_) => input.trim().to_string(),
+        _ => "".to_string() //maybe throw PANIC!
+    }
```
The `to_string()` is duplicated code here... Luckily Rust is awesome and it's expression based syntax allows this:

``` rust
 match io::stdin().read_line(&mut input) {
    Ok(_) => input.trim(),
    _ => "" //maybe throw PANIC!
}.to_string()
```

You may also use `into()` instead of `to_string()` here.

About this function in general: I would call it `read_line`. 
The function `read_line` returning an `Err` value is highly unlikely, so you could just call `unwrap()` in the Result from `read_line`. Or -- a slightly saner way -- you could loop until the `read_line` function does not error (== until you have read a valid String).

LukasKalbertodt commented:
About the two functions returning a `u8` to signal an error: It's ok to handle the error inside those functions (= tell the user what happened). But instead of returning `u8`, you should return a `bool`.

It would be more "rusty" to return some kind of `Result`, but I can't find a real reason why this would be better in this scenario. So just make the `u8` to `bool` and it should be ok.

LukasKalbertodt commented:
Your functions are pretty indented... I would suggest using a slightly different style to keep it "flatter". Instead of

``` rust
    let status = buf.read_u8();
    match status {
        Ok(st) => {
            // stuff with st
        },
        Err(_) => return 1
    }
```

I would write

``` rust
    let status = buf.read_u8();
    let st = match status {
        Err(_) => return 1,
        Ok(st) => st,
    };
    // stuff with st
```

You can always pull stuff out... More early returns:

```
if st != Cnv::GreetPkg as u8 {
    println!("Communication mismatch. Try again later.");
    return 1;
} else {
    // stuff
}
```

This is the same as

```
if st != Cnv::GreetPkg as u8 {
    println!("Communication mismatch. Try again later.");
    return 1;
} 
// stuff
```

Think about using this style. Makes it less complicated to comply with the 80 char rule.

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,217 @@
+                    SizeLimit::Bounded(1024));
+                match greet {
+                    Ok(gr) => {
+                        greeting = Greeting::make_greeting(gr.protocol_version,
```
Why is `greeting` declared 5 lines above? Why not say `let greeting = ...` here?

SvantjeJung committed:
Fixed code regarding several PullRequest comments

Fixed all available comments as best as possible. Excluding comment on
looping in read_line/cmd_read function because couldn't be fixed in the
way it was described in the comment.

SvantjeJung committed:
Fix loop reading commands from user + Add receive Ok

Changed logging on user-side from println! to info!
Need new implemented functionality from server to run client.

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,228 @@
+    let mut sl = false;
+    while !sl {
+        sl = send_login(&mut s);
+    }
```
Maybe this?

``` rust
while send_login(&mut s) {}
```

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,228 @@
+    match io::stdin().read_line(&mut input) {
+        Ok(_) => { return input.trim().into() },
+        _ => { }
+    }
```
This match expression in the `loop` needs to be indented once more.

SvantjeJung committed:
Fix logging output messages (info!) to distinct messages

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,228 @@
+    }
+    // read greeting
+    let greet = decode_from::<R, Greeting>(buf,
+        SizeLimit::Bounded(1024));
```
Those two would fit into one line, right?

LukasKalbertodt commented on src/bin/client.rs:
```diff
@@ -0,0 +1,228 @@
+    let greeting = Greeting::make_greeting(gr.protocol_version, gr.message);
+    if PROTOCOL_VERSION != greeting.protocol_version {
+        info!("Cannot communicate with server -
+                    different versions");
```
In one line?

SvantjeJung committed:
Fix code regarding style + optimizing comments
